Sources/BraintreePayPal/BTPayPalAnalytics.swift

```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+class BTPayPalAnalytics {
+    
+    static let vaultRequestStarted = "paypal:vault-tokenize:started"
+    static let checkoutRequestStarted = "paypal:checkout-tokenize:started"
+    
+    // MARK: - Tokenize Events Counted in Conversion Rates
+    static let tokenizeFailed = "paypal:tokenize:failed"
+    static let tokenizeSucceeded = "paypal:tokenize:succeeded"
+    static let browserLoginCanceled = "paypal:tokenize:browser-login:canceled"
+   
+    // MARK: - Tokenize Events Additional Detail analytic Messages
+    static let tokenizeNetworkConnectionFailed = "paypal:tokenize:network-connection:failed"
+
+    static let browserPresentationStarted = "paypal:tokenize:browser-presentation:started"
+    static let browserPresentationSucceeded = "paypal:tokenize:browser-presentation:succeeded"
+    static let browserPresentationFailed = "paypal:tokenize:browser-presentation:failed"
+    static let websessionBrowserCancel = "paypal:tokenize:websession-browser:canceled"
```

Comment:
Agreed! 

Back to the[ PPCP list,](https://paypal.sharepoint.com/:x:/r/sites/BTSDK/_layouts/15/Doc.aspx?sourcedoc=%7BAAA7C353-5744-4F21-AC7D-DF85D8605817%7D&file=Northstar%20Analytics%20Events.xlsx&action=default&mobileredirect=true&cid=1b0afccd-68ab-448f-9760-60ed14df9811) we have `browser-presentation:succeeded/failed` just for - "did the browser display or not".

Then we have `broswer-login:succeded/canceled/failed` for - "what happened with the PayPal flow? Did they successfully login and return back to app? Did they cancel? Did it fail for some other redirect reason?"

Jax - do you think those are readable?

Reply:
Yeah, I think what we have on PPCP makes sense! The ones you listed I think will make this a lot more clear.

Reply:
Sammy and I discussed that browser-login was the only way to have general "cancel" mode so we left the "browser-login:canceled" as general cancel from user selecting "cancel" from alert or from the PayPal browser. I left the "websession" word in for the specific ones because it was easier for me to distinguish those from general one.

Reply:
So to summarize, we want to have 1>` browser-login: succeeded/canceled/failed`
2> we don't distinguished where (alert or browser cancel button) it was canceled?
3>  Do we want general `tokenizeCanceled` or do we just use a `single browser-login:canceled` as a general cancel as we have now?

Reply:
> 1> browser-login: succeeded/canceled/failed

Correct

> browser-login:succeeded place would be same as tokenizeSucceeded.

I don't think necessarily, this would be fired when the PayPal flow returns to the merchant app with a success status

> browser-login:failed would be all the errors that are not from cancel in the completion block of ASWebAuthenticationSession where we also send the general tokenizeFailed.

Correct - but we should only send `tokenizeFailed` in these cases when we are calling the completion and returning. Anything that falls through to `handleBrowserSwitchReturn` should have it's tokenize event handled at that point.

> 2> we don't distinguished where (alert or browser cancel button) it was canceled?
> 3> Do we want general tokenizeCanceled or do we just use a single browser-login:canceled as a general cancel as we have now?

For cancel we want:
- `web-session-alert:canceled` - cancel for only the alert view controller 
- `browser-login:canceled` - generic cancel event for anything that's not an alert cancel (so cancel in ASWeb or cancel returned from the brower switch button as the bottom of the PayPal screen)

Reply:
When we click on the hyperlink for to cancel on the bottom of the screen, it is not caught as an error until `handleBrowserSwitchReturn`'s check for valid response. From here, it's impossible to determine whether it was from clicking on the cancel hyperlink.

Reply:
Yep - that's expected, that button is not owned by ASWeb and is owned by the screen that's displayed, so we should use the generic cancel for that case.